internal/search/query/transformer.go

```diff
@@ -10,3 +12,52 @@ func LowercaseFieldNames(nodes []Node) []Node {
 		return Parameter{Field: strings.ToLower(field), Value: value, Negated: negated}
 	})
 }
+
+// HoistOr is a heuristic that rewrites simple but possibly ambiguous queries
+// containing or-expressions. It changes certain expressions in a way that some
+// consider to be more natural. For example, the following query without
+// parentheses is interpreted as follows in the grammar:
+//
+// repo:foo a or b and c => (repo:foo a) or ((b) and (c))
+//
+// This function rewrites the above expression as follows:
+//
+// repo:foo a or b and c => repo:foo (a or b and c)
+//
+// Any number of field:value parameters may occur before and after the pattern
+// expression separated by or-operators, and these are hoisted out. The pattern
+// expression must be contiguous. If not, we want to preserve the default
+// interpretation, which corresponds more naturally to groupings with field
+// parameters, i.e.,
+//
+// repo:foo a or b or repo:bar c => (repo:foo a) or (b) or (repo:bar c)
+func HoistOr(nodes []Node) ([]Node, error) {
+	if len(nodes) != 1 {
+		return nil, fmt.Errorf("heuristic requires top-level or-expression")
+	}
+
+	expression, ok := nodes[0].(Operator)
+	if !ok || expression.Kind == And || expression.Kind == Concat {
```

Comment:
Any reason to not just check for `Or` directly?
```suggestion
	if !ok || expression.Kind != Or {
```

Reply:
Nope, just a small oversight